### DISCUSSION TITLE
Use Wireguard in kernel space if available on Android

### DISCUSSION THREAD

**Discussion by srappan** [mullvad/mullvadvpn-app#8923] [Feature requests & Ideas]
So happy to see a MullVad app on android. A feature missing from the app thats available on the official wireguard app, is if wireguard is available in the kernel, theat becomes the preferred option, as it is available in custom kernels. 
Does require root however.

**Comment by faern** [comment]
It's on our roadmap, but far in the future. Until it works on stock Android the user base is going to be tiny.

**Comment by PacoBell** [comment]
Just use the official WireGuard client if you want kernel space integration. Works fine with my custom kernel.

**Comment by lkjasdlfwe** [comment]
> Just use the official WireGuard client if you want kernel space integration. Works fine with my custom kernel.

Official Wireguard app is extremely barebones. You have to set up split tunneling and IP exclusion for each server independently. It also does not switch servers automatically if one fails, unless you turn all of them on individually, which requires you to set up split tunneling and excluding private IPs for hundreds of servers which is extremely tedious.

**Comment by monethass** [comment]
Is it possible to have this bumped up in the priority list? Google has since added the Wireguard module into Android 12's kernel in October 2020, and even backported the to kernel version 4.19, so theoretically any devices that are launched with Android 12+ since then already has the module built in, they just need root for it. Granted, it's still small compared to the non-rooted phones, but given the user base of Mullvad, it wouldn't be a surprise if the percentage of Mullvad users that have rooted their phone is non-trivial.

**Comment by veloute** [comment]
would really love to have this
